File: 2616.py
class Solution:
    def minimizeMax(self, nums: list[int], p: int) -> int:
        if p == 0:
            return 0

        nums.sort()
        # print("sorted numbers:")
        # print(nums)
        pairs = [(nums[i + 1] - nums[i], i) for i in range(len(nums) - 1)]
        # print("difference between each adjcent pairs")
        # print(pairs)
        pairs.sort()
        # print("differences sorted, with their original position")
        # print(pairs)

        m = [0] * len(nums)
        for i, (d, loc) in enumerate(pairs):
            l, r = m[loc - 1], m[loc + 1]
            m[loc] = m[loc - l] = m[loc + r] = l + r + 1
            p -= 1 & ~(l | r)
            # print("\n",m)
            # print("diff:", d,"loc:", loc, "length of left:", l, "length of right:" ,r)
            # print("remaining pairs:", p)
            if p == 0:
                return d


Solution().minimizeMax([10, 1, 2, 7, 1, 3], 2)

File: 2616.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class Solution {
  bool feasible(vector<int> &nums, int p, int max_acceptable_difference) {
    int countOfPairs = 0;
    for (int i = 0; i < nums.size() - 1;) {
      if (nums[i + 1] - nums[i] <= max_acceptable_difference) {
        countOfPairs += 1;
        i += 2;
        if (countOfPairs >= p) {
          return true;
        }
      } else {
        i += 1;
      }
    }
    return false;
  }

public:
  int minimizeMax(vector<int> &nums, int p) {
    if (p == 0) {
      return 0;
    }
    sort(nums.begin(), nums.end());
    int l = 0, r = nums[nums.size() - 1] - nums[0];
    while (l < r) {
      int m = (r - l) / 2 + l;
      if (feasible(nums, p, m)) {
        r = m;
      } else {
        l = m + 1;
      }
    }
    return l;
  }
};
int main() {
  vector<int> t1_in = {0, 5, 3, 4};
  Solution().minimizeMax(t1_in, 0);
}